pulsar-client/src/main/java/org/apache/pulsar/client/api/Producer.java

```diff
@@ -30,7 +30,7 @@
  *
  *
  */
-public interface Producer extends Closeable {
+public interface Producer<T> extends Closeable {
```

Comment:
How would one express the concept of partition key with these?

Reply:
As it stands they can't. At the moment the PartitionedProducerImpl and the ProducerImpl are hidden behind the subscribe methods which expose the Producer interface. We can either expose a PartitionedProducer or bake this functionality into the current Producer. Either way this is an important piece of functionality to call out but I think we can address this in a separate issue/pr. We should at least write up an issue because the Consumer will probably need to at least pass the partition a message came from.

Reply:
The key is currently always a string, so we don't need to have that as a template